src/Domain/Structure/Hypercube.cpp

```diff
@@ -200,6 +205,15 @@ bool HypercubeElementsIterator<ElementDim, HypercubeDim>::operator!=(
          index_ != rhs.index_;
 }
 
+template <size_t ElementDim, size_t HypercubeDim>
+bool HypercubeElementsIterator<ElementDim, HypercubeDim>::operator==(
+    const HypercubeElementsIterator& rhs) const noexcept {
+  // Implemented in terms of `operator!=` because an iterator only requires the
+  // inequality operator, and this equality operator is implemented only to
+  // complete the interface
```

Comment:
I don't understand this comment.  ("An iterator only requires the inequality operator"?  Input iterators certainly need both.)  Just do comparisons the same way they are done everywhere else.  Then you don't have to explain it.

Also, binary operators should be free functions unless they access private state.